### PR TITLE
feat(text-field): Convert JS to TypeScript

### PR DESCRIPTION
**Pull request by acdvorak**
Refs #4225

### CODE REVIEW

File: packages/mdc-textfield/types.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+export interface FoundationMapType {
+  helperText: MDCTextFieldHelperTextFoundation;
+  characterCounter: MDCTextFieldCharacterCounterFoundation;
+  leadingIcon: MDCTextFieldIconFoundation;
+  trailingIcon: MDCTextFieldIconFoundation;
+}
+
+export type RippleFactory = (el: Element, foundation: MDCRippleFoundation) => MDCRipple;
+export type LineRippleFactory = (el: Element) => MDCLineRipple;
```
Comment (acdvorak) [inline packages/mdc-textfield/types.ts]: Depends on PR #4373, PR #4374, and PR #4375.

File: packages/mdc-textfield/adapter.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+  /**
+   * @return true if the root element contains the given class name.
+   */
+  hasClass(className: string): boolean;
+
+  /**
+   * Registers an event handler on the root element for a given event.
+   */
+  registerTextFieldInteractionHandler(type: string, handler: EventListener): void;
```
Comment (kfranqueiro) [inline packages/mdc-textfield/adapter.ts]: Same comment as https://github.com/material-components/material-components-web/pull/4373#discussion_r255133145

Both here and in the icon sub-component

File: packages/mdc-textfield/types.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+import {MDCRipple, MDCRippleFoundation} from '@material/ripple';
+import {MDCTextFieldCharacterCounter} from './character-counter';
+import {MDCTextFieldCharacterCounterFoundation} from './character-counter/foundation';
+import {MDCTextFieldHelperText} from './helper-text';
+import {MDCTextFieldHelperTextFoundation} from './helper-text/foundation';
+import {MDCTextFieldIcon} from './icon';
+import {MDCTextFieldIconFoundation} from './icon/foundation';
+
+export interface NativeInputType {
```
Comment (kfranqueiro) [inline packages/mdc-textfield/types.ts]: I take it this type is only needed due to the default return of `getNativeInput_` in text field's foundation... I wonder if we can get away w/o that but I suppose this might not be the PR to find that out in.

File: packages/mdc-textfield/foundation.ts [hunk 4]
```diff
@@ -322,18 +293,17 @@ class MDCTextFieldFoundation extends MDCFoundation {
   }
 
   /**
-   * @return {boolean} If a custom validity is set, returns that value.
-   *     Otherwise, returns the result of native validity checks.
+   * Otherwise, returns the result of native validity checks.
```
Comment (kfranqueiro) [inline packages/mdc-textfield/foundation.ts]: It looks like you accidentally nuked the first line of the parameter description

File: packages/mdc-textfield/foundation.ts [hunk 5]
```diff
@@ -345,115 +315,101 @@ class MDCTextFieldFoundation extends MDCFoundation {
...
+  isDisabled(): boolean {
     return this.getNativeInput_().disabled;
   }
 
   /**
-   * @param {boolean} disabled Sets the text-field disabled or enabled.
+   * @param disabled Sets the text-field disabled or enabled.
    */
-  setDisabled(disabled) {
+  setDisabled(disabled: boolean): void {
```
Comment (kfranqueiro) [inline packages/mdc-textfield/foundation.ts]: Don't need `: void` for foundation/component methods with no return

File: packages/mdc-textfield/foundation.ts [hunk 6]
```diff
@@ -345,115 +315,101 @@ class MDCTextFieldFoundation extends MDCFoundation {
...
+   * user-supplied value
+   */
+  private isBadInput_(): boolean {
     return this.getNativeInput_().validity.badInput;
   }
 
   /**
-   * @return {boolean} The result of native validity checking
-   *     (ValidityState.valid).
+   * (ValidityState.valid).
```
Comment (kfranqueiro) [inline packages/mdc-textfield/foundation.ts]: Lost the first line of this parameter description too

File: packages/mdc-textfield/foundation.ts [hunk 7]
```diff
@@ -503,21 +455,20 @@ class MDCTextFieldFoundation extends MDCFoundation {
   }
 
   /**
-   * @return {!Element|!NativeInputType} The native text input from the
-   * host environment, or a dummy if none exists.
-   * @private
+   * host environment, or a dummy if none exists
```
Comment (kfranqueiro) [inline packages/mdc-textfield/foundation.ts]: Lost the first line of this parameter description too

File: packages/mdc-textfield/foundation.ts [hunk 8]
```diff
@@ -345,115 +315,101 @@ class MDCTextFieldFoundation extends MDCFoundation {
...
+    const maxLength = this.getNativeInput_().maxLength;
+    if (maxLength === -1) {
+      throw new Error('MDCTextFieldFoundation: Expected maxlength html property on text input or textarea.');
+    }
+
+    this.characterCounter_.setCounterValue(currentLength, maxLength);
+  }
+
+  /**
+   * user-supplied value
```
Comment (kfranqueiro) [inline packages/mdc-textfield/foundation.ts]: Lost the first line of this parameter description too

File: packages/mdc-textfield/index.ts [hunk 9]
```diff
@@ -409,21 +349,21 @@ class MDCTextField extends MDCComponent {
...
+        // tslint:disable:object-literal-sort-keys
         addClass: (className) => this.root_.classList.add(className),
         removeClass: (className) => this.root_.classList.remove(className),
         hasClass: (className) => this.root_.classList.contains(className),
         registerTextFieldInteractionHandler: (evtType, handler) => this.root_.addEventListener(evtType, handler),
         deregisterTextFieldInteractionHandler: (evtType, handler) => this.root_.removeEventListener(evtType, handler),
         registerValidationAttributeChangeHandler: (handler) => {
-          const getAttributesList = (mutationsList) => mutationsList.map((mutation) => mutation.attributeName);
+          const getAttributesList = (mutationsList: MutationRecord[]) => {
+            return mutationsList.map((mutation) => mutation.attributeName!).filter((attributeName) => attributeName);
```
Comment (kfranqueiro) [inline packages/mdc-textfield/index.ts]: Why was this changed?

File: packages/mdc-textfield/adapter.ts (hunk 2 above)
Comment (acdvorak) [reply packages/mdc-textfield/adapter.ts]: Done

File: packages/mdc-textfield/foundation.ts (hunk 4 above)
Comment (acdvorak) [reply packages/mdc-textfield/foundation.ts]: Done

File: packages/mdc-textfield/foundation.ts (hunk 6 above)
Comment (acdvorak) [reply packages/mdc-textfield/foundation.ts]: Done

File: packages/mdc-textfield/foundation.ts (hunk 5 above)
Comment (acdvorak) [reply packages/mdc-textfield/foundation.ts]: Done

File: packages/mdc-textfield/foundation.ts (hunk 8 above)
Comment (acdvorak) [reply packages/mdc-textfield/foundation.ts]: Done

File: packages/mdc-textfield/foundation.ts (hunk 7 above)
Comment (acdvorak) [reply packages/mdc-textfield/foundation.ts]: Done

File: packages/mdc-textfield/types.ts (hunk 1 above)
Comment (acdvorak) [reply packages/mdc-textfield/types.ts]: PRs have been merged

File: packages/mdc-textfield/index.ts [hunk 10]
```diff
@@ -22,87 +22,54 @@
...
-    this.outline_;
-  }
-
-  /**
-   * @param {!Element} root
-   * @return {!MDCTextField}
-   */
-  static attachTo(root) {
+import {MDCTextFieldIcon} from './icon/index';
+import {FoundationMapType} from './types';
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply packages/mdc-textfield/index.ts]: No idea... merged them into one.

File: packages/mdc-textfield/index.ts [hunk 11]
```diff
@@ -22,87 +22,54 @@
...
-   * @param {(function(!Element): !MDCFloatingLabel)=} labelFactory A function which
-   * creates a new MDCFloatingLabel.
-   * @param {(function(!Element): !MDCNotchedOutline)=} outlineFactory A function which
-   * creates a new MDCNotchedOutline.
-   */
+  // Public visibility for these properties is required by RippleCapableSurface.
+  root_!: HTMLElement; // assigned in MDCComponent constructor
+  ripple!: MDCRipple | null; // assigned in initialize()
+
+  private input_!: HTMLInputElement; // assigned in initialize()
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply packages/mdc-textfield/index.ts]: Marked all except `input_` as nullable

File: packages/mdc-textfield/index.ts [hunk 12]
```diff
@@ -153,13 +121,20 @@ class MDCTextField extends MDCComponent {
...
-          isSurfaceActive: () => this.input_[MATCHES](':active'),
-          registerInteractionHandler: (type, handler) => this.input_.addEventListener(type, handler),
-          deregisterInteractionHandler: (type, handler) => this.input_.removeEventListener(type, handler),
-        });
+      const adapter = {
+        ...MDCRipple.createAdapter(this),
+        ...({
+          // tslint:disable:object-literal-sort-keys
+          isSurfaceActive: () => ponyfill.matches(this.input_, ':active'),
+          registerInteractionHandler: (type: string, handler: EventListener) => {
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply packages/mdc-textfield/index.ts]: These ones can be removed, but the ones in `getInputAdapterMethods_()` need to remain because that method returns an anonymous object. I tried typing the methods as `Partial<MDCTextFieldAdapter>` but that blew up `tsc`.

File: packages/mdc-textfield/index.ts [hunk 13]
```diff
@@ -194,203 +169,168 @@ class MDCTextField extends MDCComponent {
...
   }
 
   /**
-   * @param {number} maxLength Sets the input element's maxLength.
+   * @param maxLength Sets the input element's maxLength.
    */
-  set maxLength(maxLength) {
-    // Chrome throws exception if maxLength is set < 0
+  set maxLength(maxLength: number) {
+    // Chrome throws exception if maxLength is se 0
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply packages/mdc-textfield/index.ts]: Done

File: packages/mdc-textfield/index.ts (hunk 9 above)
Comment (acdvorak) [reply packages/mdc-textfield/index.ts]: `handler()` expects an argument of type `string[]`, but `mutation.attributeName` can technically return `null`. Without `filter()`, `getAttributesList()` returns `(string | null)[]`.

File: packages/mdc-textfield/types.ts (hunk 3 above)
Comment (acdvorak) [reply packages/mdc-textfield/types.ts]: Correct

File: packages/mdc-textfield/adapter.ts [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+  deregisterValidationAttributeChangeHandler(observer: MutationObserver): void;
+
+  /**
+   * @return An object representing the native text input element, with a similar API shape.
+   *     We never alter the value within our code, however we do update the disabled
+   *     property, so if you choose to duck-type the return value for this method
+   *     in your implementation it's important to keep this in mind. Also note that
+   *     this method can return null, which the foundation will handle gracefully.
+   */
+  getNativeInput(): HTMLInputElement | NativeInputType | null;
```
Comment (moog16) [inline packages/mdc-textfield/adapter.ts]: I think this should just extend from `NativeInputType`, and then have that extend from `HTMLInputElement`

File: packages/mdc-textfield/foundation.ts [hunk 15]
```diff
@@ -22,127 +22,110 @@
...
+    // tslint:enable:object-literal-sort-keys
+  }
+
+  private isFocused_ = false;
+  private receivedUserInput_ = false;
+  private isValid_ = true;
+  private useNativeValidation_ = true;
+
+  private readonly inputFocusHandler_: () => void;
+  private readonly inputBlurHandler_: EventListener;
```
Comment (moog16) [inline packages/mdc-textfield/foundation.ts]: All these EventListeners should be mdc-base/SpecificEventListeners<>

File: packages/mdc-textfield/foundation.ts [hunk 16]
```diff
@@ -22,127 +22,110 @@
...
+  private readonly inputInputHandler_: EventListener;
+  private readonly setPointerXOffset_: EventListener;
+  private readonly textFieldInteractionHandler_: EventListener;
+  private readonly validationAttributeChangeHandler_: (attributesList: string[]) => void;
+  private validationObserver_!: MutationObserver; // assigned in init()
+
+  private readonly helperText_: MDCTextFieldHelperTextFoundation | undefined;
```
Comment (moog16) [inline packages/mdc-textfield/foundation.ts]: change the `| undefined` --> `helperText_?: MDCTextFieldHelperTextFoundation`, as well as the rest below

File: packages/mdc-textfield/foundation.ts [hunk 17]
```diff
@@ -246,17 +229,12 @@ class MDCTextFieldFoundation extends MDCFoundation {
...
-  setTransformOrigin(evt) {
-    let targetEvent;
-    if (evt.touches) {
-      targetEvent = evt.touches[0];
-    } else {
-      targetEvent = evt;
-    }
-    const targetClientRect = targetEvent.target.getBoundingClientRect();
-    const normalizedX = targetEvent.clientX - targetClientRect.left;
+  setTransformOrigin(evt: Event): void {
```
Comment (moog16) [inline packages/mdc-textfield/foundation.ts]: if you change the above types to SpecificEventListeners, then this won't require a cast to `as TouchEvent`

File: packages/mdc-textfield/foundation.ts [hunk 18]
```diff
@@ -503,21 +460,24 @@ class MDCTextFieldFoundation extends MDCFoundation {
...
-   * @return {!Element|!NativeInputType} The native text input from the
-   * host environment, or a dummy if none exists.
-   * @private
+   * @return The native text input from the host environment, or a dummy if none exists.
    */
-  getNativeInput_() {
-    return this.adapter_.getNativeInput() ||
-    /** @type {!NativeInputType} */ ({
-      value: '',
+  private getNativeInput_(): HTMLInputElement | NativeInputType {
```
Comment (moog16) [inline packages/mdc-textfield/foundation.ts]: This should just be `NativeInputType`

File: packages/mdc-textfield/icon/foundation.ts [hunk 19]
```diff
@@ -22,66 +22,59 @@
...
+    // tslint:enable:object-literal-sort-keys
   }
 
-  /**
-   * @param {!MDCTextFieldIconAdapter} adapter
-   */
-  constructor(adapter) {
-    super(Object.assign(MDCTextFieldIconFoundation.defaultAdapter, adapter));
+  private savedTabIndex_: string | null;
+  private readonly interactionHandler_: EventListener;
```
Comment (moog16) [inline packages/mdc-textfield/icon/foundation.ts]: Should be `SpecificEventListener<'click'> | SpecificEventListener<'keydown'>`

File: packages/mdc-textfield/foundation.ts (hunk 17 above)
Comment (acdvorak) [reply packages/mdc-textfield/foundation.ts]: We'll still need to cast it, because this param will be `evt: MouseEvent | TouchEvent`, and the two event types have different properties

File: packages/mdc-textfield/adapter.ts (hunk 14 above)
Comment (acdvorak) [reply packages/mdc-textfield/adapter.ts]: As we discussed offline, we're going to keep this as-is because we need to return a dummy object for unit tests.

File: packages/mdc-textfield/foundation.ts (hunk 15 above)
Comment (acdvorak) [reply packages/mdc-textfield/foundation.ts]: Done

File: packages/mdc-textfield/foundation.ts (hunk 16 above)
Comment (acdvorak) [reply packages/mdc-textfield/foundation.ts]: Done

File: packages/mdc-textfield/foundation.ts (hunk 18 above)
Comment (acdvorak) [reply packages/mdc-textfield/foundation.ts]: Done

File: packages/mdc-textfield/icon/foundation.ts (hunk 19 above)
Comment (acdvorak) [reply packages/mdc-textfield/icon/foundation.ts]: Done

File: test/unit/mdc-textfield/foundation.test.js [hunk 20]
```diff
@@ -403,35 +421,55 @@ test('#init does not float label if the input does not contain a value', () => {
...
+});
+
 test('#setTrailingIconContent sets the content of the trailing icon element', () => {
-  const {foundation, trailingIcon} = setupTest();
+  const {foundation, trailingIcon} = setupTest({useTrailingIcon: true});
   foundation.setTrailingIconContent('foo');
   td.verify(trailingIcon.setContent('foo'));
 });
 
+test('#setTrailingIconContent sets does nothing if element is not present', () => {
```
Comment (kfranqueiro) [inline test/unit/mdc-textfield/foundation.test.js]: remove "sets"

File: test/unit/mdc-textfield/foundation.test.js [hunk 21]
```diff
@@ -403,35 +421,55 @@ test('#init does not float label if the input does not contain a value', () => {
...
+});
+
 test('#setLeadingIconContent sets the content of the leading icon element', () => {
-  const {foundation, leadingIcon} = setupTest();
+  const {foundation, leadingIcon} = setupTest({useLeadingIcon: true});
   foundation.setLeadingIconContent('foo');
   td.verify(leadingIcon.setContent('foo'));
 });
 
+test('#setLeadingIconContent sets does nothing if element is not present', () => {
```
Comment (kfranqueiro) [inline test/unit/mdc-textfield/foundation.test.js]: Remove "sets"

File: test/unit/mdc-textfield/mdc-text-field-character-counter-foundation.test.js [hunk 22]
```diff
@@ -53,7 +53,9 @@ test('#setContent sets the content of the character counter element', () => {
 });
 
 test('#setContent current length does not exceed character count limit', () => {
-  const {foundation, mockAdapter} = setupTest();
+  const foundation = new MDCTextFieldCharacterCounterFoundation();
+  const adapter = foundation.adapter_;
+  adapter.setContent = td.func('setContent');
```
Comment (kfranqueiro) [inline test/unit/mdc-textfield/mdc-text-field-character-counter-foundation.test.js]: I think I asked this on one of the tab PRs and I'm asking it again here: why are we suddenly using a partially-mocked adapter? What does this improve in code coverage? Is there something else we should be doing to improve it instead?

File: packages/mdc-textfield/foundation.ts [hunk 23]
```diff
@@ -22,127 +22,113 @@
...
+    };
+    // tslint:enable:object-literal-sort-keys
+  }
+
+  private isFocused_ = false;
+  private receivedUserInput_ = false;
+  private isValid_ = true;
+  private useNativeValidation_ = true;
+
+  private readonly inputFocusHandler_: () => void;
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply packages/mdc-textfield/foundation.ts]: Because we call it in `init()` further below, and we don't have an `Event` argument to pass it:

```ts
  init() {
    if (this.adapter_.isFocused()) {
      // `SpecificEventListener` requires an `Event` argument.
      // We could conceivably pass `null` or `document.createEvent('FocusEvent')`,
      // but that feels wrong.
      this.inputFocusHandler_();
    }
  }
```

File: packages/mdc-textfield/foundation.ts [hunk 24]
```diff
@@ -345,115 +322,102 @@ class MDCTextFieldFoundation extends MDCFoundation {
...
   }
 
   /**
-   * @return {boolean} The result of native validity checking
-   *     (ValidityState.valid).
+   * @return True if the Text Field input fails in converting the user-supplied value.
    */
-  isNativeInputValid_() {
+  private isBadInput_(): boolean {
+    // The badInput property is not supported in IE 11 💩.
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply packages/mdc-textfield/foundation.ts]: Hah I guess I've gotten used to it and didn't even notice... Personally I think it's hilarious and fitting, but I can remove it if you really want 😛

File: packages/mdc-textfield/helper-text/index.ts [hunk 25]
```diff
@@ -22,45 +22,31 @@
...
       addClass: (className) => this.root_.classList.add(className),
       removeClass: (className) => this.root_.classList.remove(className),
       hasClass: (className) => this.root_.classList.contains(className),
       setAttr: (attr, value) => this.root_.setAttribute(attr, value),
       removeAttr: (attr) => this.root_.removeAttribute(attr),
-      setContent: (content) => {
-        this.root_.textContent = content;
-      },
-    })));
+      setContent: (content) => this.root_.textContent = content,
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply packages/mdc-textfield/helper-text/index.ts]: [You are technically correct](https://i.imgur.com/PpV0evc.png). Does it matter tho? What would you suggest doing instead?

[the remark this answers is not in the document]
Comment (kfranqueiro) [reply packages/mdc-textfield/helper-text/index.ts]: Retain the `{...}` around the ones that shouldn't return anything. (I'm surprised it compiles.)

File: packages/mdc-textfield/icon/foundation.ts [hunk 26]
```diff
@@ -22,66 +22,61 @@
...
+  private savedTabIndex_: string | null;
+  private readonly interactionHandler_: SpecificEventListener<InteractionEventType>;
 
-    /** @private {string?} */
-    this.savedTabIndex_ = null;
+  constructor(adapter?: Partial<MDCTextFieldIconAdapter>) {
+    super({...MDCTextFieldIconFoundation.defaultAdapter, ...adapter});
 
-    /** @private {function(!Event): undefined} */
+    this.savedTabIndex_ = null;
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply packages/mdc-textfield/icon/foundation.ts]: Done

File: packages/mdc-textfield/index.ts [hunk 27]
```diff
@@ -0,0 +1,451 @@
...
+
+  // Public visibility for these properties is required by RippleCapableSurface.
+  root_!: HTMLElement; // assigned in MDCComponent constructor
+  ripple!: MDCRipple | null; // assigned in initialize()
+
+  // The only required sub-element.
+  private input_!: HTMLInputElement; // assigned in initialize()
+
+  // Optional sub-elements.
+  private characterCounter_!: MDCTextFieldCharacterCounter | null; // assigned in initialize()
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply packages/mdc-textfield/index.ts]: Using `null` just made the code a little cleaner by condensing the `if` statements down to ternaries. That's the only reason I did it 😄

File: packages/mdc-textfield/index.ts [hunk 28]
```diff
@@ -0,0 +1,451 @@
...
+
+    const isTextArea = this.root_.classList.contains(cssClasses.TEXTAREA);
+    const isOutlined = this.root_.classList.contains(cssClasses.OUTLINED);
+    this.ripple = (isTextArea || isOutlined) ? null : rippleFactory(this.root_, new MDCRippleFoundation({
+      ...MDCRipple.createAdapter(this),
+      ...({
+        // tslint:disable:object-literal-sort-keys
+        isSurfaceActive: () => ponyfill.matches(this.input_, ':active'),
+        registerInteractionHandler: (evtType, handler) => {
+          return this.input_.addEventListener(evtType, handler);
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply packages/mdc-textfield/index.ts]: Removed

File: test/unit/mdc-textfield/foundation.test.js (hunk 20 above)
Comment (acdvorak) [reply test/unit/mdc-textfield/foundation.test.js]: Done

File: test/unit/mdc-textfield/foundation.test.js (hunk 21 above)
Comment (acdvorak) [reply test/unit/mdc-textfield/foundation.test.js]: Done

File: packages/mdc-textfield/helper-text/index.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (acdvorak) [reply packages/mdc-textfield/helper-text/index.ts]: Done

File: test/unit/mdc-textfield/mdc-text-field-character-counter-foundation.test.js (hunk 22 above)
Comment (acdvorak) [reply test/unit/mdc-textfield/mdc-text-field-character-counter-foundation.test.js]: Istanbul was complaining that we're not testing
`new MDCTextFieldCharacterCounterFoundation()` _without_ an argument. I tried adding `/* istanbul ignore next */` comments everywhere, but nothing worked (I'm guessing it's a quirk of transpiling and source maps).

The only thing that made Istanbul happy was testing the no-argument constructor directly like this.

File: packages/mdc-textfield/index.ts (hunk 27 above)
[the remark this answers is not in the document]
Comment (kfranqueiro) [reply packages/mdc-textfield/index.ts]: I guess since they're all private it doesn't really matter either way. I'm not sure if there's any perception that one is preferable than the other, typing-wise, other than `?` existing as a more concise representation of `| undefined`.

File: packages/mdc-textfield/types.ts [hunk 29]
```diff
@@ -0,0 +1,53 @@
...
+import {MDCRippleFoundation} from '@material/ripple/foundation';
+import {MDCRipple} from '@material/ripple/index';
+import {MDCTextFieldCharacterCounterFoundation} from './character-counter/foundation';
+import {MDCTextFieldCharacterCounter} from './character-counter/index';
+import {MDCTextFieldHelperTextFoundation} from './helper-text/foundation';
+import {MDCTextFieldHelperText} from './helper-text/index';
+import {MDCTextFieldIconFoundation} from './icon/foundation';
+import {MDCTextFieldIcon} from './icon/index';
+
+export type NativeInputElement = Pick<HTMLInputElement, 'disabled' | 'maxLength' | 'type' | 'value'> & {
```
Comment (moog16) [inline packages/mdc-textfield/types.ts]: NICE

File: test/unit/mdc-textfield/foundation.test.js [hunk 30]
```diff
@@ -56,32 +56,37 @@ test('defaultAdapter returns a complete adapter implementation', () => {
...
-const setupTest = () => {
+const setupTest = ({
+  useCharacterCounter = false,
+  useHelperText = false,
+  useLeadingIcon = false,
+  useTrailingIcon = false,
+} = {}) => {
   const mockAdapter = td.object(MDCTextFieldFoundation.defaultAdapter);
-  const helperText = td.object({
+  const helperText = useHelperText === true ? td.object({
```
Comment (moog16) [inline test/unit/mdc-textfield/foundation.test.js]: I think this can just be `useHelperText ?td...`

File: test/unit/mdc-textfield/foundation.test.js [hunk 31]
```diff
@@ -56,32 +56,37 @@ test('defaultAdapter returns a complete adapter implementation', () => {
...
     setContent: () => {},
     showToScreenReader: () => {},
     setValidity: () => {},
-  });
-  const characterCounter = td.object({
+  }) : undefined;
+  const characterCounter = useCharacterCounter === true ? td.object({
```
Comment (moog16) [inline test/unit/mdc-textfield/foundation.test.js]: same with these....

File: packages/mdc-textfield/types.ts (hunk 29 above)
Comment (acdvorak) [reply packages/mdc-textfield/types.ts]: Thanks for the suggestion! 🎉 🌮

File: test/unit/mdc-textfield/foundation.test.js (hunk 30 above)
Comment (acdvorak) [reply test/unit/mdc-textfield/foundation.test.js]: Oh yeah, duh... Fixed

File: test/unit/mdc-textfield/foundation.test.js (hunk 31 above)
Comment (acdvorak) [reply test/unit/mdc-textfield/foundation.test.js]: Done

File: test/unit/mdc-textfield/mdc-text-field-character-counter-foundation.test.js (hunk 22 above)
Comment (acdvorak) [reply test/unit/mdc-textfield/mdc-text-field-character-counter-foundation.test.js]: Reverted these changes and added a separate test for Istanbul code coverage
